Add error handler for DenyAccessAuthenticator

lexcao opened the pull request:
Closes #13368

These changes introduce a new error handler for building error based on the media type.
- It should create error form response when it is valid HTML request
- Otherwise, it could create error response with JSON

[open question]
* there is not support message bundle for JsonErrorHandler, should I add it as well?

@mposolda 
<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

pedroigor commented on services/src/main/java/org/keycloak/authentication/authenticators/access/DenyAccessAuthenticator.java:
```diff
@@ -46,9 +46,12 @@ public void authenticate(AuthenticationFlowContext context) {
                 .orElse(Messages.ACCESS_DENIED);
 
         context.getEvent().error(Errors.ACCESS_DENIED);
-        Response challenge = context.form()
```
Isn't possible to keep the current construct/API and be smarter about how the response should be created based on the media type when actually calling `createErrorPage`?

lexcao replied:
Hi @pedroigor 
Sounds a good idea.
Let me try it.

pedroigor commented:
@lexcao Looks like failures are related. Would you mind checking them?

lexcao commented:
Hi @pedroigor 
I just fixed the failed test, thanks.
The error was  NPE in `flowPath` called by `getActionUrl` from `context.form()`

```
23:26:53,742 ERROR [org.keycloak.services.error.KeycloakErrorHandler] Uncaught server error
java.lang.IllegalArgumentException: RESTEASY003715: path was null
	at org.jboss.resteasy.specimpl.ResteasyUriBuilderImpl.path(ResteasyUriBuilderImpl.java:355)
	at org.keycloak.authentication.AuthenticationProcessor$Result.getActionUrl(AuthenticationProcessor.java:566)
	at org.keycloak.authentication.AuthenticationProcessor$Result.form(AuthenticationProcessor.java:540)
	at org.keycloak.authentication.authenticators.access.DenyAccessAuthenticator.authenticate(DenyAccessAuthenticator.java:49)
	at org.keycloak.authentication.DefaultAuthenticationFlow.processSingleFlowExecutionModel(DefaultAuthenticationFlow.java:445)
	at org.keycloak.authentication.DefaultAuthenticationFlow.processFlow(DefaultAuthenticationFlow.java:249)
....
```

Since there is no `flowPath` in the token endpoint, is maybe an undefined behavior.
I just hard code (see the PR) to make it with value and seek for your suggestion.

or should we be able to handle the `flowPath` in the `DenyAccessAuthenticator`?
* I do not want to handle `flowPath` in an authenticator

pedroigor commented:
> or should we be able to handle the `flowPath` in the `DenyAccessAuthenticator`?
> 
> * I do not want to handle `flowPath` in an authenticator

TBH, I need to check. I'll run your PR locally and look at the failing tests to see how it goes.

lexcao commented:
Thanks.
The failing test is `configureDirectGrantFlowWithDenyAccess` if you remove this line of code
```
// services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java
.setFlowPath("token")
```

pedroigor commented on services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java:
```diff
@@ -678,6 +678,7 @@ public Response resourceOwnerPasswordCredentialsGrant() {
         AuthenticationProcessor processor = new AuthenticationProcessor();
         processor.setAuthenticationSession(authSession)
                 .setFlowId(flowId)
+                .setFlowPath("token")
```
Looking at https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/authentication/AuthenticationProcessor.java#L544, we should be doing something very similar as you are doing in the `FreemarkerLoginFormsProvider`. If the request is a backend request we don't really need to handle forms.

@mposolda wdyt?

pedroigor reviewed (commented):
@lexcao Sorry for the long delay. Only a single comment about how we should deal with backend requests when processing forms.

Once we sort it out, LGTM.

pedroigor reviewed (commented):
@lexcao The failures seem related. Could you check it out?

Talked with @mposolda and the change you did to workaround the flow path issue is OK to move forward.

Thanks.

pedroigor commented:
@lexcao Btw, you also need to sign the commit by setting the `-s` option when creating te commit.

lexcao commented:
Hi @pedroigor 
I have updated the PR, maybe we can merge it if the CI is all green.

pedroigor reviewed (commented):
@lexcao Failure seems related. See https://github.com/keycloak/keycloak/actions/runs/7448885531/job/20264259312?pr=22006#step:3:22425.

lexcao commented:
@pedroigor 
Thanks, let me try again.

lexcao commented:
This changes break the tests for SAML POST binding mode like `org.keycloak.testsuite.broker.KcSamlBrokerTest#loginInResponseNoRelayState`.

```
Expected: response body matches a string containing "Unexpected error when authenticating with identity provider"
     but: was <HttpResponseProxy{...} with entity {"error":"identityProviderUnexpectedErrorMessage"}
```

These are requests with POST method and asserting the response body text.
But here we just respond raw error messages with JSON.

I come up with some ways to handle this,
1. Fix the broken test with JSON response assertions, which would break the current response of SAML.
2. Handle the message bundle if the tests assertions failed on the error message, which may be out of scope of this PR.
3. Responding the error only in the `DenyAccessAuthenticator`, which keep the scope.

I am not sure we could break the SAML error response, or the JSON is the correct way?
I prefer the last method.

WDYT? @pedroigor

pedroigor commented:
> I am not sure we could break the SAML error response, or the JSON is the correct way?

Yeah, we can't break such things here. 

> These are requests with POST method and asserting the response body text.
But here we just respond raw error messages with JSON.

Looks like the main issue here is that we default to returning JSON if no accept header is set. I think we should default to return HTML.

IIRC, the behavior we are adding here is more related to backend requests so for these cases we can enforce the accept header being set if you want a proper content-type as a response.

lexcao commented:
@pedroigor 
Good idea, let me try it.

lexcao committed:
Error handle for the Json request in createErrorPage

Closes #13368

These changes introduce a new error handler for building error based on the media type.
- It should create error form response when it is valid HTML request
- It could create error response with JSON if content type matches

Signed-off-by: Lex Cao <lexcao@foxmail.com>

pedroigor commented:
@lexcao Now the test failure is the opposite :) What we expect. We need to update the `OAuthClient` to set the appropriate header when making back-end requests ...

pedroigor commented:
@lexcao I see ... I think `403` makes more sense too it looks more like an authorization than an authentication check. @mposolda wdyt?

mabartos commented on services/src/main/java/org/keycloak/forms/login/freemarker/FreeMarkerLoginFormsProvider.java:
```diff
@@ -29,11 +29,8 @@
-import jakarta.ws.rs.core.UriBuilder;
-import jakarta.ws.rs.core.UriInfo;
+
+import jakarta.ws.rs.core.*;
```
@lexcao Please, don't use the star imports.

lexcao replied:
Thanks, let me fix it.

lexcao commented:
Hi @pedroigor 
I updated the PR with handling the 403 for the `UserSessionLimitsAuthenticator` as well, please help review.

pedroigor reviewed (approved):
@lexcao LGTM, thanks.

mposolda commented on services/src/main/java/org/keycloak/utils/MediaTypeMatcher.java:
```diff
@@ -1,16 +1,24 @@
+        return isAcceptMediaType(headers, MediaType.APPLICATION_JSON_TYPE);
+    }
+
+    private static boolean isAcceptMediaType(HttpHeaders headers, MediaType textHtmlType) {
```
minor: renaming the latter variable to `mediaType` might be better (not a blocker for this PR unless there are other points to address).

lexcao replied:
Thanks for the suggestion. I just forgot this one.
I will take care of the param name next time.

mposolda reviewed (approved):
@lexcao Nice work! Added one very minor comment inline, but not a blocker (unless there are other comments).
